Instructor: edit session: 'distribute points' questions: support 'some variation required' option #8449

joanneong opened the pull request:
Fixes #8449 

**Outline of Solution**

I have added support for some variation on whether givers can distribute the same number of points to all/some recipients/options.

Here's a GIF demonstration:

![constsumvariation](https://user-images.githubusercontent.com/24493463/36905839-53ed790a-1e6f-11e8-890b-e3f2039aa7ce.gif)

Behaviour summary: 
1)	Default (checkbox unchecked): the giver can give any number of points to any number of recipients/options 
2)	ALL uneven (checkbox checked): the giver must give a different number of points to all recipients/options
3)	SOME uneven (checkbox checked, selected option changed): the giver must give a different number of points to at least some recipients/options

 I have also added tests for (i) testing input validation on the student’s end and (ii) testing changing options on the instructor’s end.

However, while I was adding tests for the instructor in `InstructorFeedbackSubmitPageUiTest.java`, I observed something intriguing – if I were to fill the response text boxes for the new constant sum question with some uneven distribution in `testSubmitAction()`, under `edit existing response`, then an option will not be displayed for all the constant sum questions during the testing. I am not sure why this is happening, but for the moment I have worked around it by filling the response boxes later under `testConstSumSubmitAction()` instead.

P/S Yet to update: `instructorHelp.jsp`, which I intend to update last after these changes are confirmed.

P/P/S There’s a gap when `Every option` is selected since it is a shorter option than `At least some options`…But I think making the box change size to accommodate the option length would also be very odd. Alternatively, I think this is also a possibility:

![image](https://user-images.githubusercontent.com/24493463/36905447-5428a4d6-1e6e-11e8-994e-4910d782d086.png)

which also has a gap but the gap is not mid sentence. However, this would look different from other questions like MCQ/MSQ questions where you can select where to generate options from:

![image](https://user-images.githubusercontent.com/24493463/36905470-65924a9c-1e6e-11e8-9b2f-97317449807f.png)

What do you think?

P/P/P/S This question also suffers from [warning about deleting responses even if changes are not destructive](https://github.com/TEAMMATES/teammates/issues/7678).

damithc commented:
Good start. I'm OK with both UIs. If the second one is used, may be name it as `How points are to be distributed:`?

wenmogu commented on src/main/java/teammates/common/datatransfer/questions/FeedbackConstantSumDistributePointsType.java:
```diff
@@ -0,0 +1,23 @@
+public enum FeedbackConstantSumDistributePointsType {
+    DISTRIBUTE_ALL_UNEVENLY("All options"),
+    DISTRIBUTE_SOME_UNEVENLY("At least some options"),
+    NONE("None");
```
i like the use of enum :D for strings like above, idk if u can put them into Const, so its easier to edit...

wenmogu replied:
I like the use of enum:D but i think maybe u can put these strings into Const as they will be reused in quite some diff places?

wenmogu commented on src/main/webapp/dev/js/common/questionConstSum.js:
```diff
@@ -70,10 +70,28 @@ function removeConstSumOption(index, questionNum) {
+                $(`#constSumDistributePointsSelect-${questionNum}`).prop('value'));
+    } else {
+        $(`#constSumDistributePointsSelect-${questionNum}`).prop('disabled', true);
+        $(`#constSumDistributePointsOptions-${questionNum}`).attr('value', 'None');
```
for this part i think you can afford to use a few more lines to extract some variables and give them proper names to make the code more comprehensive. And on a random note can i ask y for some parts u use `prop(value)` to get the value, instead of `attr(value)`?

joanneong replied:
1. Can I ask what you mean by "extract some variables and give them proper names"? Here, I am only referencing two items - `constSumDistributePointsSelect` and `constSumDistributePointsOptions`. I cannot immediately see how else I can further simplify this (unless you mean the naming)...

2. Admittedly the division between `prop`  and `attr` is somewhat nebulous. However, I did this mainly because I drew inspiration from the format for MCQ/MSQ questions. Besides, I have also read that:

>`attr()` gives you the value of element as it was defined in the html on page load. It is always recommended to use `prop()` to get values of elements which is modified via javascript/jquery , as it gives you the original value of an element's current state. 

so I thought that the MCQ/MSQ way of doing this was fine.

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -491,7 +491,9 @@ function updateConstSumMessageQn(qnNum) {
-    const forceUnevenDistribution = $(`#constSumUnevenDistribution-${qnNum}`).val() === 'true';
+    const forceUnevenDistribution = $(`#constSumDistributePointsOptions-${qnNum}`).val() !== 'None';
+    const forceAllUnevenDistribution = $(`#constSumDistributePointsOptions-${qnNum}`).val() === 'All options';
+    const forceSomeUnevenDistribution = $(`#constSumDistributePointsOptions-${qnNum}`).val() === 'At least some options';
```
And for these strings here, I think you can put them in const.js

joanneong commented:
@wenmogu 

For (1) I feel like it should not be directly implemented in this PR (especially since the original system does not have this). Instead, another issue can be opened so that others can also give their inputs and decide whether this change is necessary.

For (2) I will spend some time looking into it next week :/

joanneong commented:
@wenmogu 

Update: I traced the issue and realised it was a bug in the logic that I implemented (rather than any storage issue). I have updated and tested, and now things work as per expected.

Please check if it's ready for review by a senior developer, thanks :)

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -491,7 +492,14 @@ function updateConstSumMessageQn(qnNum) {
-    const forceUnevenDistribution = $(`#constSumUnevenDistribution-${qnNum}`).val() === 'true';
+
+    const constSumDistributePointsOptionSelected = $(`#constSumDistributePointsOptions-${qnNum}`).val();
+    const forceUnevenDistribution = constSumDistributePointsOptionSelected
```
personally i find this variable confusing. Since the only place it is used is at `updateConstSumMessageQn` and the way it is used is `!forceUnevenDistribution`, which leads to a double-negation (one in the  definition one in the usage), I think you can change it into `noForceUnevenDistribution = constSumDistributePointsOptionSelected === ParamsNames.FEEDBACK_QUESTION_CONSTSUMNOUNEVENDISTRIBUTION;` and at its usage change it into `forceUnevenDistribution`.

wenmogu commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -525,7 +534,7 @@ function updateConstSumMessageQn(qnNum) {
             messageElement.removeClass('text-color-green');
         } else if (remainingPoints === 0) {
-            if (!forceUnevenDistribution || allUnique) {
+            if (!forceUnevenDistribution || allUnique || (forceSomeUnevenDistribution && !allSame)) {
```
I noticed that the condition u added corresponds to the case of "distributing some points unevenly". I like this combination of the conditions, it specifies the case (forceSomeUnevenDistribution) and the constraint(!allSame). I think you can do the same thing to the other two conditions i.e. `!forceUnvevenDistribution` and `allUnique`

joanneong commented:
@wkurniawan07 Reminder to take a look at this, thank you!

whipermr5 commented:
How is existing data interpreted?

joanneong commented:
Thanks @whipermr5 for catching this - it was indeed something that I failed to consider when I first started work on this PR :P

**Outline of changes made (Update)**

In the original design, the checkbox being checked means "All options" are to be distributed unevenly. To ensure that my version is backwards compatible, I have:
(i) changed the identifier of the checkbox back to the original identifier
(ii) changed the logic slightly such that if the checkbox is found to be checked, but without a corresponding distribution option ("None", "At least some options" or "All options") saved along with it, it will be treated as "All options"

There is also double checking for when there need not be any uneven distribution - either the checkbox is found to be unchecked, or the distribution option selected has the value "None".

I have tested it manually as well by switching between the master branch and this branch, but if there are still any issues do alert me!

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackConstantSumQuestionDetails.java:
```diff
@@ -82,7 +84,11 @@ public boolean extractQuestionDetails(
+
+        String distributePointsOption =
+                HttpRequestHelper.getValueFromParamMap(requestParameters,
+                                                       Const.ParamsNames.FEEDBACK_QUESTION_CONSTSUMDISTRIBUTEPOINTSOPTIONS);
```
Use 8-space indentation. The function call can be combined with the previous line.

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackConstantSumQuestionDetails.java:
```diff
@@ -82,7 +84,11 @@ public boolean extractQuestionDetails(
         String forceUnevenDistributionString =
                 HttpRequestHelper.getValueFromParamMap(requestParameters,
-                                                       Const.ParamsNames.FEEDBACK_QUESTION_CONSTSUMDISTRIBUTEUNEVENLY);
+                                                        Const.ParamsNames.FEEDBACK_QUESTION_CONSTSUMDISTRIBUTEUNEVENLY);
```
Same here.

joanneong replied:
Do you mean to do something like this:

```
String forceUnevenDistributionString = HttpRequestHelper.getValueFromParamMap(requestParameters,
        Const.ParamsNames.FEEBBACK_QUESTION_CONSTSUMDISTRIBUTEUNEVENLY);
```

The reason why I did that was because of the code that was already present prior to my PR:

![image](https://user-images.githubusercontent.com/24493463/38535923-ddbf6956-3cb8-11e8-9a52-131d83ea5eef.png)

But I can change all of them if this is better coding style.

whipermr5 replied:
Yup. Oh, I see. Yes; can change all of them.

whipermr5 commented:
@wkurniawan07 Agreed.

whipermr5 commented:
@joanneong You are free to merge your own PR :) Just remember to follow the naming conventions.

joanneong commented:
@whipermr5 Do I have to add an `e.x` label? I did not review this PR so...

P/S I checked the past commits and some had messages (summary of changes) in the body of their commit message (after squashing and merging) but others do not. Is it more standard to not have any summary in the body of the commit message?
